hal/src/main/native/athena/SerialPort.cpp

```diff
@@ -1,174 +1,500 @@
+  // NO OP currently
 }
 
-int32_t HAL_GetSerialBytesReceived(HAL_SerialPort port, int32_t* status) {
-  int32_t bytes = 0;
-
-  *status = viGetAttribute(portHandles[port], VI_ATTR_ASRL_AVAIL_NUM, &bytes);
-  if (*status > 0) *status = 0;
+int32_t HAL_GetSerialBytesReceived(HAL_SerialPortHandle handle,
+                                   int32_t* status) {
+  auto port = serialPortHandles->Get(handle);
+  if (!port) {
+    *status = HAL_HANDLE_ERROR;
+    return -1;
+  }
+  int bytes = 0;
+  int err = ioctl(port->portId, FIONREAD, &bytes);
+  if (err < 0) {
+    *status = errno;
+  }
   return bytes;
 }
 
-int32_t HAL_ReadSerial(HAL_SerialPort port, char* buffer, int32_t count,
+int32_t HAL_ReadSerial(HAL_SerialPortHandle handle, char* buffer, int32_t count,
                        int32_t* status) {
-  uint32_t retCount = 0;
-
-  *status =
-      viRead(portHandles[port], (ViPBuf)buffer, count, (ViPUInt32)&retCount);
+  // Don't do anything if 0 bytes were requested
+  if (count == 0) return 0;
 
-  if (*status == VI_ERROR_IO || *status == VI_ERROR_ASRL_OVERRUN ||
-      *status == VI_ERROR_ASRL_FRAMING || *status == VI_ERROR_ASRL_PARITY) {
-    int32_t localStatus = 0;
-    HAL_ClearSerial(port, &localStatus);
+  auto port = serialPortHandles->Get(handle);
+  if (!port) {
+    *status = HAL_HANDLE_ERROR;
+    return -1;
   }
 
-  if (*status == VI_ERROR_TMO || *status > 0) *status = 0;
-  return static_cast<int32_t>(retCount);
+  int n = 0, loc = 0;
+  char buf = '\0';
+  std::memset(buffer, '\0', count);
+  *status = 0;
+
+  do {
+    n = read(port->portId, &buf, 1);
+    if (n == 1) {
+      buffer[loc] = buf;
+      loc++;
+      // If buffer is full, return
+      if (loc == count) {
+        return loc;
+      }
+      // If terminating, and termination was hit return;
+      if (port->termination && buf == port->terminationChar) {
+        return loc;
+      }
+    } else if (n == -1) {
+      // ERROR
+      *status = errno;
+      return loc;
+    } else {
+      // If nothing read, timeout
+      return loc;
+    }
+  } while (true);
 }
 
-int32_t HAL_WriteSerial(HAL_SerialPort port, const char* buffer, int32_t count,
-                        int32_t* status) {
-  uint32_t retCount = 0;
-
-  *status =
-      viWrite(portHandles[port], (ViPBuf)buffer, count, (ViPUInt32)&retCount);
-
-  if (*status > 0) *status = 0;
-  return static_cast<int32_t>(retCount);
-}
+int32_t HAL_WriteSerial(HAL_SerialPortHandle handle, const char* buffer,
+                        int32_t count, int32_t* status) {
+  auto port = serialPortHandles->Get(handle);
+  if (!port) {
+    *status = HAL_HANDLE_ERROR;
+    return -1;
+  }
 
-void HAL_FlushSerial(HAL_SerialPort port, int32_t* status) {
-  *status = viFlush(portHandles[port], VI_WRITE_BUF);
-  if (*status > 0) *status = 0;
+  int written = 0, spot = 0;
+  do {
+    written = write(port->portId, buffer + spot, count - spot);
+    if (written > 0) {
+      spot += written;
+    } else if (written < 0) {
```

Comment:
Would it be worth writing this with less branches? Like so:
```c
if (written < 0) {
  *status = errno;
  return spot;
}
spot += written;
```
To me this feels less confusing at least.

Reply:
Fixed